elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +265,144 @@ private Object fetchRelationship(Environment context) {
         }
     }
 
+    /**
+     * handle UPSERT operation
+     * @param context environment encapsulating graphQL's request environment
+     * @return list of persistent resource objects
+     */
+    private Object upsertObjects(Environment context) {
+        /* sanity check for id and data argument w UPSERT */
+        if(context.ids.isPresent()) {
+            throw new BadRequestException("UPSERT must not include ids");
+        }
+
+        if(!context.data.isPresent()) {
+            throw new BadRequestException("UPSERT must include data argument");
+        }
+
+        Set<PersistentResource> upsertedObjects = new HashSet<>();
+        EntityDictionary dictionary = context.requestScope.getDictionary();
+
+        for(Map<String, Object> input : context.data.get()) {
+            if(context.isRoot()) {
+                /* get internal reference idFieldName, 'id' in this case */
+                String idFieldName = dictionary.getIdFieldName(dictionary.getEntityClass(context.field.getName()));
+
+                upsertedObjects.add(upsertObject(idFieldName, input, context.requestScope,
+                        context.field, context.outputType));
+            } else {
+                String idFieldName = dictionary.getIdFieldName(context.parentResource.getResourceClass());
+
+                upsertedObjects.add(upsertObject(idFieldName, input, dictionary, context.parentResource,
+                        context.field, context.outputType, context.requestScope));
+            }
+        }
+        return upsertedObjects;
+    }
+
+    /**
+     * handle upsert on non root-level entities
+     * @param idFieldName id field name
+     * @param input input data object
+     * @param dictionary entity dictionary
+     * @param parentSource parent persistent resource object
+     * @param field graphql-java field type
+     * @param outputType graphql-java output type
+     * @return list/set of persistent resource objects
+     */
+    private PersistentResource upsertObject(String idFieldName, Map<String, Object> input, EntityDictionary dictionary,
+                                PersistentResource parentSource, Field field, GraphQLType outputType,
+                                RequestScope requestScope) {
+        /* fetch id valued fed to 'data' argument */
+        Set<String> ids = input.entrySet().stream().filter(entry -> idFieldName.equalsIgnoreCase(entry.getKey()))
+                .map(entry -> (String) entry.getValue())
+                .collect(Collectors.toSet());
+
+        try {
+            Object fetchEntries = fetchObject(dictionary, parentSource, field, outputType.getClass().getName());
+            if(fetchEntries == null) {
+                return createObject(input, outputType, requestScope, Optional.of(ids), field, parentSource);
+            } else {
+                return updateObject(); //TODO: implement update
+            }
+        } catch (BadRequestException e) {
```

Comment:
Close the try block right after fetch so it is explicit this exception handling only applies for the fetch.